design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,272 @@
+Finally, CRDs are not a breaking change for Rook users since they are already in use today.
+
+### Controller Options
+
+The controllers are the entities that will perform orchestration and deployment tasks to ensure the users desired state is made a reality within their cluster.
+There are a few options for deploying and hosting the controllers that will perform this work, as explained in the following sections.
+
+#### In-proc Controllers
+
+Multiple controllers could be deployed within a single process.
+For instance, Rook could run one controller per domain of expertise, e.g., ceph-controller, minio-controller, etc.
+Controllers would all watch the same custom resource types via a `SharedInformer` and respond to events via a `WorkQueue` for efficiency and reduced burden on core apiserver.
+Even though all controllers are watching, only the applicable controller responds to and handles an event.
+For example, the user runs `kubectl create cluster.yaml` to create a `cluster.rook.io` instance that has Ceph specific properties.
+All controllers will receive the created event via the `SharedInformer`, but only the Ceph controller will queue and handle it.
+We can consider only loading the controllers the user specifically asks for, perhaps via an environment variable in `rook-operator.yaml`.
+
+Note that this architecture can be used with either API aggregation or CRDs.
+
+##### PROS
+
+* Potentially easier developer burden for new storage backends since there is no need to create a new deployment to host their controller.
+* Less resource burden on K8s cluster since watchers/caches are being shared.
+* Easier migration path to API aggregation in the future, if CRDs usage is continued now.
+
+##### CONS
+
+* All controllers must use same base image since they are all running in the same process.
+* If a controller needs to access a backend specific tool then they will have to schedule a job that invokes the appropriate image.  This is similar to exec’ing a new process but at the cluster level.
+  * We have very limited need for this so far with the Ceph implementation (e.g., creating the initial crush map).
+  * Note this only applies to the controller, not to the backend's daemons.  Those daemons will be running a backend specific image and can directly `exec` to their tools.
+
+#### Operator pods
+
+Each storage backend could have their own operator pod that hosts only that backend's controller, e.g., `ceph-operator.yaml`, `minio-operator.yaml`, etc.
+The user would decide which operators to deploy based on what storage they want to use.
+Each operator pod would watch the same custom resource types with their own individual watchers.
+
+##### PROS
+
+* Each operator can use their own image, meaning they have direct access (through `exec`) to their backend specific tools.
```

Comment:
another pro is that you get resource isolation between backends. One flaky backend won't affect another.

Reply:
Generally, each storage cluster will probably prefer to deal with storage failures (including loss of nodes) itself. It will gladly welcome more resources when Kubernetes get around to assigning them, but continued operation of the storage service will generally not be dependent on relauch a new pod. The only exception I can imagine is if a storage service is a Metadata Server service that uses an underlying  SAN service for persistence. In that case, it may be acceptable to wait for Kubernetes to relaunch the MDS to take ownership of the same SAN resources. But if it is providing persistence itself it should already know how to deal with the loss of any single storage device, and at scale with loss of multiple storage devices.